src/system_identity/generator.rs

```diff
@@ -0,0 +1,78 @@
+use std::fmt;
+
+use thiserror::Error;
+
+use crate::key::creator::Creator as KeyCreator;
+
+use super::{
+    identity_creator::{L1IdentityCreator, L2IdentityCreator},
+    SystemIdentity,
+};
+
+/// This type is responsible for generating a System Identity and its associated key pair.
+pub struct L2SystemIdentityGenerator<K, I>
+where
+    K: KeyCreator,
+    I: L2IdentityCreator,
+{
+    pub key_creator: K,
+    pub iam_client: I,
+}
+
+impl<K, I> L2SystemIdentityGenerator<K, I>
+where
+    K: KeyCreator,
+    I: L2IdentityCreator,
+{
+    pub fn generate(&self) -> Result<SystemIdentity, SystemIdentityGenerationError<K, I>> {
+        let pub_key = self
+            .key_creator
+            .create()
+            .map_err(SystemIdentityGenerationError::KeyPairCreator)?;
+
+        self.iam_client
+            .create_l2_system_identity(pub_key.as_slice())
+            .map_err(SystemIdentityGenerationError::IAMClient)
+    }
+}
+
+pub struct L1SystemIdentityGenerator<I: L1IdentityCreator> {
+    pub iam_client: I,
+}
+
+impl<I: L1IdentityCreator> L1SystemIdentityGenerator<I> {
+    pub fn generate(&self) -> Result<SystemIdentity, I::Error> {
+        self.iam_client.create_l1_system_identity()
+    }
+}
+
```

Comment:
I really like how explicit is now with the L1 and L2 generators. It makes me doubt about having `L1SystemIdentity` and `L2SystemIdentity` maybe as traits, and `SystemIdentity` implementing both. But it might be too much, or if not we could add the abstraction later.